sw/device/tests/entropy_src_fw_ovr_test.c

```diff
@@ -0,0 +1,127 @@
+  uint32_t read_count = 0;
+
+  // TODO: Remove this limit. Entropy source should block if there is no entropy
+  // available in FW override mode.
+  const uint32_t kMaxReadCount = 12;
+
+  while (dif_entropy_src_avail(entropy) == kDifEntropySrcOk) {
+    CHECK(dif_entropy_src_read(entropy, &entropy_bits) == kDifEntropySrcOk);
+    CHECK(entropy_bits != 0);
+    read_count++;
+    if (read_count >= kMaxReadCount) {
+      break;
+    }
+  }
+  LOG_INFO("Flushed %d entropy words.", read_count);
+}
+
+/**
+ * Configures the entropy source module in firmware override mode.
+ *
+ * Output is routed to firmware, and the fw_override mode is enabled to get data
+ * post-health tests and before the pre conditioner block.
+ *
+ * @param entropy An entropy source instance.
+ */
+static void entropy_with_fw_override_enable(dif_entropy_src_t *entropy) {
+  const dif_entropy_src_config_t config = {
+      .mode = kDifEntropySrcModeLfsr,
+      .tests =
+          {
+              [kDifEntropySrcTestRepCount] = false,
+              [kDifEntropySrcTestAdaptiveProportion] = false,
+              [kDifEntropySrcTestBucket] = false,
+              [kDifEntropySrcTestMarkov] = false,
+              [kDifEntropySrcTestMailbox] = false,
+              [kDifEntropySrcTestVendorSpecific] = false,
+          },
+      .reset_health_test_registers = false,
+      .single_bit_mode = kDifEntropySrcSingleBitModeDisabled,
+      .route_to_firmware = true,
+      .sample_rate = 2,
+      .lfsr_seed = 2,
+      .fw_override =
+          {
+              .enable = true,
+              .entropy_insert_enable = true,
+              .buffer_threshold = kEntropyFifoBufferSize,
+          },
+  };
+  CHECK(dif_entropy_src_configure(entropy, config) == kDifEntropySrcOk);
+}
+
+/**
+ * Test the firmware override path.
+ *
+ * This tests disconnects the observation buffer from the entropy src
+ * pre-conditioner block to read the raw entropy data after the hardware
+ * health tests. It then feeds the data back into the conditioner block until
+ * there is data available in the output FIFO.
+ *
+ * @param entropy An Entropy handle.
+ */
+void test_firmware_override(dif_entropy_src_t *entropy) {
+  CHECK(dif_entropy_src_disable(entropy) == kDifEntropySrcOk);
+  entropy_with_fw_override_enable(entropy);
+  entropy_data_flush(entropy);
+
+  // Read data from the obeservation and write it back into the pre conditioner
+  // until there is data available in the output buffer.
+  uint32_t buf[kEntropyFifoBufferSize] = {0};
+  uint32_t word_count = 0;
+  do {
+    CHECK(dif_entropy_src_fifo_read(entropy, buf, kEntropyFifoBufferSize) ==
+          kDifEntropySrcOk);
+    for (size_t i = 0; i < kEntropyFifoBufferSize; ++i) {
+      CHECK(buf[i] != 0);
+    }
+    CHECK(dif_entropy_src_fifo_write(entropy, buf, kEntropyFifoBufferSize) ==
+          kDifEntropySrcOk);
+    word_count += kEntropyFifoBufferSize;
+  } while (dif_entropy_src_avail(entropy) == kDifEntropySrcDataUnAvailable);
```

Comment:
Likewise, if using the globally defined error codes in `sw/device/lib/dif/dif_base.h`, this will likely become `kDifUnavailable`.